client/beefy/src/worker.rs

```diff
@@ -500,19 +500,27 @@ where
 
 			self.on_demand_justifications.cancel_requests_older_than(block_num);
 
-			if let Err(e) = self.backend.append_justification(
-				BlockId::Number(block_num),
-				(BEEFY_ENGINE_ID, finality_proof.encode()),
-			) {
-				error!(target: "beefy", "🥩 Error {:?} on appending justification: {:?}", e, finality_proof);
-			}
+			let maybe_block_hash =
+				self.backend.blockchain().expect_block_hash_from_id(&BlockId::Number(block_num));
+
+			match maybe_block_hash {
+				Ok(hash) => {
+					if let Err(err) = self
+						.backend
+						.append_justification(&hash, (BEEFY_ENGINE_ID, finality_proof.encode()))
+					{
+						error!(target: "beefy", "🥩 Error {:?} on appending justification: {:?}", err, finality_proof);
+					};
 
-			self.backend.blockchain().hash(block_num).ok().flatten().map(|hash| {
-				self.links
-					.to_rpc_best_block_sender
-					.notify(|| Ok::<_, ()>(hash))
-					.expect("forwards closure result; the closure always returns Ok; qed.")
-			});
+					self.links
+						.to_rpc_best_block_sender
+						.notify(|| Ok::<_, ()>(hash))
+						.expect("forwards closure result; the closure always returns Ok; qed.");
+				},
+				Err(err) => {
+					error!(target: "beefy", "🥩 Error {:?} hash not know for given block_num: {:?}", err, finality_proof);
+				},
+			};
```

Comment:
This block could be reworked to output a single error message. Anyone has preferences about that?

Reply:
would prefer single error message here, but current version is ok too.

Reply:
single message in: [4930349](https://github.com/paritytech/substrate/pull/12551/commits/4930349a6e97a7d96564da1cb95459ca942d5e79)